src/player.cpp

```diff
@@ -11158,205 +11158,276 @@ void player::gunmod_add( item &gun, item &mod )
+        reasons.push_back( string_format( _( "Your %s is not good reading material." ),
+                                          book.tname().c_str() ) );
+        return nullptr;
     }
 
-    if (!it->is_book()) {
-        add_msg(m_info, _("Your %s is not good reading material."),
-        it->tname().c_str());
-        return false;
+    // Check for conditions that immediately disqualify the player from reading:
+    const vehicle *veh = g->m.veh_at( pos() );
+    if( veh != nullptr && veh->player_in_control( *this ) ) {
+        reasons.push_back( _( "It's a bad idea to read while driving!" ) );
+        return nullptr;
+    }
+    auto type = book.type->book.get();
+    if( type->fun <= 0 && !has_morale_to_read() && has_identified( book.typeId() ) ) {
+        // Low morale still permits skimming
+        reasons.push_back( _( "What's the point of studying?  (Your morale is too low!)" ) );
+        return nullptr;
+    }
+    const skill_id &skill = type->skill;
+    const auto &skill_level = get_skill_level( skill );
+    if( skill && skill_level < type->req ) {
+        reasons.push_back( string_format( _( "You don't know enough about %s to understand the jargon!" ),
+                                          skill.obj().name().c_str() ) );
+        return nullptr;
+    }
+
+    // Check for conditions tha disqualify us only if no NPCs can read to us
+    if( type->intel > 0 && has_trait( "ILLITERATE" ) ) {
+        reasons.push_back( _( "You're illiterate!" ) );
+    } else if( has_trait( "HYPEROPIC" ) && !is_wearing( "glasses_reading" ) &&
+               !is_wearing( "glasses_bifocal" ) && !has_effect( effect_contacts ) ) {
+        reasons.push_back( _( "Your eyes won't focus without reading glasses." ) );
+    } else if( fine_detail_vision_mod() > 4 ) {
+        // Too dark to read only applies if the player can read to himself
+        reasons.push_back( _( "It's too dark to read!" ) );
+        return nullptr;
+    } else {
+        reader = this;
     }
 
     //Check for NPCs to read for you, negates Illiterate and Far Sighted
-    int assistants = 0;
-    for( auto &elem : g->active_npc ) {
-        if (rl_dist( elem->pos(), g->u.pos() ) < PICKUP_RANGE && elem->is_friend()){
-            ///\EFFECT_INT_NPC allows NPCs to read harder books for you
-            if ((elem->int_cur+1) >= it->type->book->intel)
-                assistants++;
-        }
+    //The fastest-reading NPC is chosen
+    if( !reader ) {
+        int time_taken = INT_MAX;
+        if( is_deaf() ) {
+            reasons.push_back( _( "Maybe someone could read that to you, but you're deaf!" ) );
+            return nullptr;
+        }
+        for( const npc *elem : g->active_npc ) {
+            // Check for disqualifying factors:
+            if( !elem->is_friend() ) {
+                ; // Don't consider non-friends
+            } else if( type->intel > 0 && elem->has_trait( "ILLITERATE" ) ) {
+                if( sees( *elem ) ) {
```

Comment:
You check `sees` in every block. You should move it out to the top, as a separate `if`, before any of those.